dialogue-core/src/main/java/com/palantir/dialogue/core/ContentDecodingChannel.java

```diff
@@ -53,21 +54,43 @@ final class ContentDecodingChannel implements EndpointChannel {
     private static final String CONTENT_ENCODING = "content-encoding";
     private static final String CONTENT_LENGTH = "content-length";
     private static final String GZIP = "gzip";
+    private static final String PREFER_COMPRESSED_RESPONSE_TAG = "prefer-compressed-response";
 
     private final EndpointChannel delegate;
+    private final boolean sendAcceptGzip;
 
-    ContentDecodingChannel(EndpointChannel delegate) {
+    private ContentDecodingChannel(EndpointChannel delegate, boolean sendAcceptGzip) {
         this.delegate = Preconditions.checkNotNull(delegate, "Channel is required");
+        this.sendAcceptGzip = sendAcceptGzip;
+    }
+
+    static EndpointChannel create(Config cf, EndpointChannel delegate, Endpoint endpoint) {
+        boolean sendAcceptGzip = shouldSendAcceptGzip(cf, endpoint);
+        return new ContentDecodingChannel(delegate, sendAcceptGzip);
+    }
+
+    private static boolean shouldSendAcceptGzip(Config cf, Endpoint endpoint) {
+        // If the override tag has been configured, always request gzipped responses.
+        if (endpoint.tags().contains(PREFER_COMPRESSED_RESPONSE_TAG)) {
+            return true;
+        }
+        // In mesh mode or environments which appear to be within an environment,
+        // prefer not to request compressed responses. This heuristic assumes response
+        // compression should not be used in a service mesh, nor when load balancing
+        // is handled by the client.
+        return cf.mesh() == MeshMode.DEFAULT_NO_MESH && cf.clientConf().uris().size() == 1;
     }
 
     @Override
     public ListenableFuture<Response> execute(Request request) {
-        Request augmentedRequest = acceptEncoding(request);
+        Request augmentedRequest = acceptEncoding(request, sendAcceptGzip);
+        // In cases where gzip is not expected, we continue to handle gzipped responses to avoid abrupt failures
+        // against servers which hard-code 'Content-Encoding: gzip' responses without checking request headers.
         return DialogueFutures.transform(delegate.execute(augmentedRequest), ContentDecodingChannel::decompress);
     }
 
-    private static Request acceptEncoding(Request request) {
-        if (request.headerParams().containsKey(ACCEPT_ENCODING)) {
+    private static Request acceptEncoding(Request request, boolean sendAcceptGzip) {
+        if (!sendAcceptGzip || request.headerParams().containsKey(ACCEPT_ENCODING)) {
```

Comment:
Seems like `sendAcceptGzip` is now some extra state that `ContentDecodingChannel` holds. Curious what you think of adding it as a param in `toString` like `"ContentDecodingChannel{delegate=" + delegate + " sendAcceptGzip=" + sendAcceptGzip + '}'`?

Reply:
Good idea, updated!